src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request).Add(id)].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        protected override void PostPersist(ApplicationEngine engine)
+        {
+            base.PostPersist(engine);
+            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                //Filter the response transactions
+                OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+                if (response is null) continue;
+
+                //Remove the request from storage
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+                int index = (int)(response.Id % (ulong)nodes.Length);
+                UInt160 account = Contract.CreateSignatureRedeemScript(nodes[index]).ToScriptHash();
+                GAS.Mint(engine, account, OracleRequestPrice);
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForRepsonse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForRepsonse <= 0)
```

Comment:
Should we check `url` here? We only support `http` and `https` at present.

Reply:
It should be checked by the Oracle nodes, not by contract.

Reply:
Add `if OracleNodes.Empty throw new InvalidOperation()` ?

Reply:
If there is no oracle nodes, the requests can be stored and be proceeded once the oracle nodes are available.

Reply:
if `0 < gasForRepsonse < MinResponseTxFee(OracleResponse.Success = false) ` , what will happen?

Reply:
> If there is no oracle nodes, the requests can be stored and be proceeded once the oracle nodes are available.

I think when there is no oracle nodes, it means oracle service is unavailable. If we store the request, users must wait for uncertainty.

Reply:
What is `MinResponseTxFee`?

Reply:
I mean when `gasForRepsonse` is insufficient to pay the response transaction, what will happen? Will oracle nodes still get reward?  If not, when oracle nodes restart, they will try to handle the request again.

Reply:
Rewards and deletions occur in `PostPersist`.

Reply:
But when `gasForRepsonse` is insufficient, there is no response tx, then we can't give rewards and delete.

Reply:
You are right. `gasForRepsonse` should pay the network fee.

Reply:
@Tommo-L I believe c9bb097 should fix this issue.

Reply:
see https://github.com/neo-project/neo/pull/1738#commitcomment-40740420